File: apps/mobile/lib/services/service_area_geometry_codec.dart
class ServiceAreaGeometryCodec {
  static const encoding = 'map-parts-v1';

  static Map<String, dynamic> decodePreferences(Map<String, dynamic> input) {
    final output = Map<String, dynamic>.from(input)..remove('geometryEncoding');
    output['areas'] = (input['areas'] as List? ?? const [])
        .whereType<Map>()
        .map((area) => decodeArea(Map<String, dynamic>.from(area)))
        .toList(growable: false);
    return output;
  }

  static Map<String, dynamic> decodeArea(Map<String, dynamic> area) {
    final output = Map<String, dynamic>.from(area)..remove('geometryEncoding');
    final geometry = _points(area['geometry']);
    final parts = (area['geometryParts'] as List? ?? const [])
        .map(_points)
        .where((part) => part.length >= 3)
        .toList(growable: false);
    output['geometry'] = geometry;
    output['geometryParts'] = parts.isNotEmpty
        ? parts
        : geometry.length >= 3
        ? [geometry]
        : <List<Map<String, dynamic>>>[];
    return output;
  }

  static List<Map<String, dynamic>> _points(Object? value) {
    final raw = value is Map ? value['points'] : value;
    return (raw as List? ?? const [])
        .whereType<Map>()
        .map((point) => Map<String, dynamic>.from(point))
        .where((point) => point['latitude'] is num && point['longitude'] is num)
        .toList(growable: false);
  }
}

File: functions-admin-ops/service_area_geometry_codec.js
"use strict";

const GEOMETRY_ENCODING = "map-parts-v1";
function point(value) {
  const latitude = Number(value?.latitude); const longitude = Number(value?.longitude);
  if (!Number.isFinite(latitude) || !Number.isFinite(longitude) || latitude < -90 ||
      latitude > 90 || longitude < -180 || longitude > 180) return null;
  return {latitude, longitude};
}
function points(value) { return Array.isArray(value) ? value.map(point).filter(Boolean) : []; }
function runtimeParts(area) {
  const geometry = points(area?.geometry?.points || area?.geometry);
  const parts = Array.isArray(area?.geometryParts) ? area.geometryParts.map((part) =>
    points(part?.points || part)).filter((part) => part.length >= 3) : [];
  return {geometry, geometryParts: parts.length ? parts : (geometry.length >= 3 ? [geometry] : [])};
}
function encodeServiceAreaGeometryForFirestore(area) {
  const normalized = runtimeParts(area);
  return {...area, geometryEncoding: GEOMETRY_ENCODING, geometry: {points: normalized.geometry},
    geometryParts: normalized.geometryParts.map((part) => ({points: part}))};
}
function decodeServiceAreaGeometryFromFirestore(area) {
  if (!area || typeof area !== "object") return area;
  const normalized = runtimeParts(area);
  const decoded = {...area, geometry: normalized.geometry, geometryParts: normalized.geometryParts};
  delete decoded.geometryEncoding;
  return decoded;
}
function encodeDiscoveryPreferencesForFirestore(preferences) {
  return {...preferences, geometryEncoding: GEOMETRY_ENCODING,
    areas: Array.isArray(preferences?.areas) ? preferences.areas
      .map(encodeServiceAreaGeometryForFirestore) : []};
}
function decodeDiscoveryPreferencesFromFirestore(preferences) {
  if (!preferences || typeof preferences !== "object") return preferences;
  const decoded = {...preferences, areas: Array.isArray(preferences.areas) ? preferences.areas
    .map(decodeServiceAreaGeometryFromFirestore) : []};
  delete decoded.geometryEncoding;
  return decoded;
}
function containsDirectNestedArray(value) {
  if (Array.isArray(value)) return value.some(Array.isArray) || value.some(containsDirectNestedArray);
  return value && typeof value === "object" ? Object.values(value).some(containsDirectNestedArray) : false;
}
module.exports = {GEOMETRY_ENCODING, encodeServiceAreaGeometryForFirestore,
  decodeServiceAreaGeometryFromFirestore, encodeDiscoveryPreferencesForFirestore,
  decodeDiscoveryPreferencesFromFirestore, containsDirectNestedArray};
